### PR TITLE
feature/PT-4/local_storage

### PR DESCRIPTION
**Pull request by dungeon-slave**
add pop-up with info about connection
implement menu saving in local storage
add textScale feature
add textScale saving in local storage
add settings item

https://github.com/dungeon-slave/food_app/assets/70040363/08c0fb34-f624-46c3-ad97-1be86314b302

### CODE REVIEW

File: core_ui/lib/src/theme/app_constants.dart [hunk 1]
```diff
@@ -18,4 +27,9 @@ final class AppConstants {
 
   static const String cartCheckout = 'Checkout';
   static const String cartTotal = 'Total';
+
+  static const String connectionLoss = 'Disconnected!';
+  static const String connectionRestored = 'Connected!';
+
+  static const List<double> textScales = [0.8, 1.0, 1.2];
```
Comment (ibondarovich) [inline core_ui/lib/src/theme/app_constants.dart:34]: use enum

File: data/lib/providers/remote/firebase_provider.dart [hunk 2]
```diff
@@ -4,7 +4,7 @@ import 'package:data/entities/dish_entity/dish_entity.dart';
 import 'package:data/entities/dish_type_enity/dish_type_entity.dart';
 
 class FirebaseProvider {
-  Future<List<DishTypeEntity>> fetchDishes() async {
+  Future<List<DishTypeEntity>> getMenu() async {
```
Comment (ibondarovich) [inline data/lib/providers/remote/firebase_provider.dart:7]: use fetch

File: data/pubspec.lock [hunk 3]
```diff
@@ -461,6 +461,14 @@ packages:
       url: "https://pub.dev"
     source: hosted
     version: "4.0.2"
+  internet_connection_checker:
```
Comment (ibondarovich) [inline data/pubspec.lock:464]: remove locks

File: domain/lib/repositories/dishes_repository.dart [hunk 4]
```diff
@@ -1,5 +1,6 @@
 import 'package:domain/models/dishes_items/dish_type_model.dart';
 
 abstract class DishesRepository {
-  Future<List<DishTypeModel>> fetchDishes();
+  Future<List<DishTypeModel>> getMenu();
```
Comment (ibondarovich) [inline domain/lib/repositories/dishes_repository.dart:4]: fetch

File: domain/lib/repositories/text_scale_repository.dart [hunk 5]
```diff
@@ -0,0 +1,4 @@
+abstract class TextScaleRepository {
+  Future<void> saveTextScale(double textScale);
+  double getTextScale();
```
Comment (ibondarovich) [inline domain/lib/repositories/text_scale_repository.dart:3]: same

File: domain/lib/usecase/dishes/fetch_dishes_usecase.dart [hunk 6]
```diff
@@ -2,13 +2,16 @@ import 'package:domain/domain.dart';
...
 
-  const FetchDishesUsecase(this._dishesRepository);
+  const FetchDishesUsecase({
+    required DishesRepository dishesRepository,
+  }) : _dishesRepository = dishesRepository;
 
   @override
   Future<List<DishTypeModel>> execute(NoParams empty) async {
-    return _dishesRepository.fetchDishes();
+    return _dishesRepository.getMenu();
```
Comment (ibondarovich) [inline domain/lib/usecase/dishes/fetch_dishes_usecase.dart:15]: same

File: domain/lib/usecase/text_scale/get_text_scale_usecase.dart [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+
+class GetTextScaleUseCase implements UseCase<NoParams, double> {
+  final TextScaleRepository _textScaleRepository;
+
+  const GetTextScaleUseCase({
+    required TextScaleRepository textScaleRepository,
+  }) : _textScaleRepository = textScaleRepository;
+
+  @override
+  double execute(NoParams input) => _textScaleRepository.getTextScale();
```
Comment (ibondarovich) [inline domain/lib/usecase/text_scale/get_text_scale_usecase.dart:12]: fetch

File: domain/lib/usecase/text_scale/set_text_scale_usecase.dart [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+
+class SetTextScaleUseCase implements AsyncUseCase<double, void> {
+  final TextScaleRepository _textScaleRepository;
+
+  const SetTextScaleUseCase({
+    required TextScaleRepository textScaleRepository,
+  }) : _textScaleRepository = textScaleRepository;
+
+  @override
+  Future<void> execute(double input) async => await _textScaleRepository.saveTextScale(input);
```
Comment (ibondarovich) [inline domain/lib/usecase/text_scale/set_text_scale_usecase.dart:12]: formatting

File: features/home_screen/lib/src/bloc/home_screen_bloc.dart [hunk 9]
```diff
@@ -10,14 +11,53 @@ class HomeScreenBloc extends Bloc<HomeScreenEvent, HomeScreenState> {
...
+          ChangeConnectionEvent(
+            isConnected: event == InternetConnectionStatus.connected,
+          ),
+        );
+      },
+    );
   }
 
   void _changeCartCount(
       ChangeCartCountEvent event, Emitter<HomeScreenState> emit) {
```
Comment (ibondarovich) [inline features/home_screen/lib/src/bloc/home_screen_bloc.dart:36]: ,

File: features/home_screen/lib/src/bloc/home_screen_bloc.dart [hunk 10]
```diff
@@ -10,14 +11,53 @@ class HomeScreenBloc extends Bloc<HomeScreenEvent, HomeScreenState> {
...
     emit(
-      HomeScreenState(count: _getCartCountUseCase.execute(const NoParams())),
+      state.copyWith(
+        count: _getCartCountUseCase.execute(const NoParams()),
+      ),
     );
   }
+
+  void _changeMessageVisibility(
+      ShowMessageEvent event, Emitter<HomeScreenState> emit) {
```
Comment (ibondarovich) [inline features/home_screen/lib/src/bloc/home_screen_bloc.dart:45]: ,

File: features/home_screen/lib/src/bloc/home_screen_bloc.dart [hunk 11]
```diff
@@ -10,14 +11,53 @@ class HomeScreenBloc extends Bloc<HomeScreenEvent, HomeScreenState> {
...
+    emit(
+      state.copyWith(
+        isVisibleMessage: false,
+      ),
+    );
+  }
+
+  void _changeConnectionStatus(
+      ChangeConnectionEvent event, Emitter<HomeScreenState> emit) {
```
Comment (ibondarovich) [inline features/home_screen/lib/src/bloc/home_screen_bloc.dart:54]: ,

File: features/settings/lib/src/bloc/settings/settings_bloc_bloc.dart [hunk 12]
```diff
@@ -10,26 +12,72 @@ class SettingsBloc extends Bloc<ThemeEvent, SettingsState> {
...
+        _urlService = urlService,
+        super(SettingsState(
+          isDark: true,
+          textScale: AppConstants.textScales[1],
+        )) {
     on<SetThemeEvent>(_setTheme);
     on<GetThemeEvent>(_getTheme);
+    on<SetTextScaleEvent>(_setTextScale);
+    on<GetTextScaleEvent>(_getTextScale);
+    on<OpenLinkEvent>(_openLink);
```
Comment (ibondarovich) [inline features/settings/lib/src/bloc/settings/settings_bloc_bloc.dart:39]: space

File: features/settings/lib/src/bloc/settings/settings_bloc_bloc.dart [hunk 13]
```diff
@@ -10,26 +12,72 @@ class SettingsBloc extends Bloc<ThemeEvent, SettingsState> {
   final SetThemeUseCase _setThemeUseCase;
   final GetThemeUseCase _getThemeUseCase;
```
Comment (ibondarovich) [inline features/settings/lib/src/bloc/settings/settings_bloc_bloc.dart:14]: remove space

File: data/lib/providers/local/hive_provider.dart [hunk 14]
```diff
@@ -48,10 +42,12 @@ class HiveProvider {
     await _cartBox.put(item.dishEntity.id, item);
   }
 
-  Future<void> clearCart() async => await _cartBox.clear();
+  Future<void> saveTextScale(double textScale) async => await _textScaleBox.put(_HiveKeys.textScaleKey, textScale);
```
Comment (ibondarovich) [inline data/lib/providers/local/hive_provider.dart:45]: formatting

File: data/lib/providers/local/hive_provider.dart [hunk 15]
```diff
@@ -48,10 +42,12 @@ class HiveProvider {
...
 
   bool getTheme() => _themeBox.get(_HiveKeys.themeKey) ?? true;
 
+  double getTextScale() => _textScaleBox.get(_HiveKeys.textScaleKey) ?? AppConstants.textScales[1];
```
Comment (ibondarovich) [inline data/lib/providers/local/hive_provider.dart:49]: formatting

File: data/lib/providers/local/hive_provider.dart [hunk 16]
```diff
@@ -1,37 +1,31 @@
 import 'package:core/core.dart';
+import 'package:core_ui/core_ui.dart';
 import 'package:data/entities/cart_item_entity/cart_item_entity.dart';
+import 'package:data/entities/dish_type_enity/dish_type_entity.dart';
 
 class HiveProvider {
   late final Box<bool> _themeBox;
```
Comment (ibondarovich) [inline data/lib/providers/local/hive_provider.dart:7]: remove late?

**Review by ibondarovich — CHANGES_REQUESTED** [review]

File: data/pubspec.lock (hunk 3 above)
Comment (dungeon-slave) [reply data/pubspec.lock:464]: I removed them in next PR

File: features/settings/lib/src/bloc/settings/settings_bloc_bloc.dart (hunk 13 above)
Comment (dungeon-slave) [reply features/settings/lib/src/bloc/settings/settings_bloc_bloc.dart:14]: Enter

File: data/lib/providers/local/hive_provider.dart (hunk 16 above)
Comment (dungeon-slave) [reply data/lib/providers/local/hive_provider.dart:7]: I dont think that

File: core_ui/lib/src/theme/app_constants.dart (hunk 1 above)
Comment (dungeon-slave) [reply core_ui/lib/src/theme/app_constants.dart:34]: (optional)

Comment (Alexey-Lebedzeu) [reply core_ui/lib/src/theme/app_constants.dart:34]: enum

**Pull request merged**